plugins/plugin-machine/che-plugin-machine-ext-server/src/main/java/org/eclipse/che/ide/ext/machine/server/ssh/KeysInjector.java

```diff
@@ -54,72 +75,100 @@ public void start() {
         new EventSubscriber<MachineStatusEvent>() {
           @Override
           public void onEvent(MachineStatusEvent event) {
-            if (event.getEventType() == MachineStatus.RUNNING) {
-              /*final Instance machine;
+            if (event.getEventType() != MachineStatus.RUNNING) {
```

Comment:
I think you should rework it to ServerStatusEvent. Because when Machine is running it doesn't mean that exec agent is RUNNING. See https://github.com/eclipse/che/issues/8056

Reply:
Its a problem than. Because startup of sinlge agent is not guarantee that all other agents is started successfully. Startup still may be failed later. So only way it to do that later, on whole `workspace` startup.

Reply:
I'd say that it is not related to listening server statuses or machine statuses. Because if one machine is started it is not guaranteed that all machines will be started.
But I don't think that it is a problem. You inject ssh key as soon as possible. Can you elaborate what is the problem of failing a workspace when ssh key is already injected?

Reply:
because:
1. This is not logical (inject SSH key each time when exec agent started? how they are related?)
2. This is useless action when startup failed.
3. This is highly error risky (you submitiing a task to workspace which is begin to shutdown and state is changing you can loss your actual state of command etc)

Reply:
Makes sense. But 2 and 3 points are also related MachineStatusEvent too. So `WorkspaceStatus` is OK for me.